src/components/EditProduct.tsx

```diff
@@ -0,0 +1,349 @@
+    statuses: string[];
+    duplicateName: boolean;
+}
+
+function EditProduct({ match }: IProps) {
+    const { allowed, apiClient, redirect } = useContext(ApplicationContext);
+    const [productLoaded, setProductLoaded] = useState(false);
+    const [confirmationDialogOpen, setConfirmationDialogOpen] = useState(false);
+    const [confirmationDialogAction, setConfirmationDialogAction] = useState<() => void>(() => null);
+    const [cancelDialogAction, setCancelDialogAction] = useState<() => void>(() => null);
+    const [confirmationDialogQuestion, setConfirmationDialogQuestion] = useState("");
+    const [leavePage, setLeavePage] = useState(true);
+    const [errors, setErrors] = useState<Partial<Record<keyof ProductData, boolean>>>({});
+    const [required, setRequired] = useState<(keyof ProductData)[]>([
+        "name",
+        "description",
+        "status",
+        "product_type",
+        "tag",
+    ]);
+    const [initial, setInitial] = useState(true);
+    // const [readOnly, setReadOnly] = useState(false)
+    const [processing, setProcessing] = useState(false);
+    const [duplicateName, setDuplicateName] = useState(false);
+    const [products, setProducts] = useState<Product[]>([]);
+    const [statuses, setStatuses] = useState<string[]>([]);
+
+    const [product, setProduct] = useState<Product>({
+        create_subscription_workflow_key: "",
+        created_at: 0,
+        description: "",
+        end_date: 0,
+        fixed_inputs: [],
+        modify_subscription_workflow_key: "",
+        name: "",
+        product_blocks: [],
+        product_id: "",
+        product_type: "",
+        status: "",
+        tag: "",
+        terminate_subscription_workflow_key: "",
+        workflows: [],
+    });
+
+    // class EditProduct extends React.Component<IProps, IState> {
+    //     state: IState = {
+    //         confirmationDialogOpen: false,
+    //         confirmationDialogAction: () => this.setState({ confirmationDialogOpen: false }),
+    //         cancelDialogAction: () => this.context.redirect("/metadata/products"),
+    //         confirmationDialogQuestion: "",
+    //         leavePage: true,
+    //         errors: {},
+    //         required: ["name", "description", "status", "product_type", "tag"],
+    //         initial: true,
+    //         readOnly: false,
+    //         processing: false,
+    //         productBlocks: [],
+    //         products: [],
+    //         workflows: [],
+    //         tags: [],
+    //         types: [],
+    //         statuses: [],
+    //         duplicateName: false,
+    //     };
+    debugger;
+    useEffect(() => {
+        const product_id = match?.params.id;
+        if (product_id) {
+            apiClient.productById(product_id).then((res: Product) => {
+                console.log("if products");
+                setProduct(res);
+                setProductLoaded(true);
+            });
+            apiClient.products().then((res: Product[]) => {
```

Comment:
Is a call to all products needed? The old construction seemed to work without it.